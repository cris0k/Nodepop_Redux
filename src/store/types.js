export const AUTH_LOGIN_REQUEST = 'AUTH_LOGIN_REQUEST';
export const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
export const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';

export const AUTH_LOGOUT = 'AUTH_LOGOUT';

export const ADVERTS_LOADED_REQUEST = 'ADVERTS_LOADED_REQUEST';
export const ADVERTS_LOADED_SUCCESS = 'ADVERTS_LOADED_SUCCESS';
export const ADVERTS_LOADED_FAILURE = 'ADVERTS_LOADED_FAILURE';

export const DETAILS_LOADED_REQUEST = 'DETAILS_LOADED_REQUEST';
export const DETAILS_LOADED_SUCCESS = 'DETAILS_LOADED_SUCCESS';
export const DETAILS_LOADED_FAILURE = 'DETAILS_LOADED_FAILURE';

export const ADVERT_CREATED_REQUEST = 'ADVERT_CREATED_REQUEST';
export const ADVERT_CREATED_SUCCESS = 'ADVERT_CREATED_SUCCESS';
export const ADVERT_CREATED_FAILURE = 'ADVERT_CREATED_FAILURE';


export const UI_RESET_ERROR = 'UI_RESET_ERROR'