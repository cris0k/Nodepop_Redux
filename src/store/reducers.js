import { 
  AUTH_LOGIN_SUCCESS, 
  AUTH_LOGOUT, 
  ADVERTS_LOADED_SUCCESS, 
  UI_RESET_ERROR,
  DETAILS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS} 
  from './types';

export const defaultState = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  ui: {
    isLoading: false,
    error: null
   }
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED_SUCCESS) {
    return { areLoaded:true, data: action.payload};
  }
  if (action.type === DETAILS_LOADED_SUCCESS){
    return {...state, data: [action.payload]}
  }
  if (action.type === ADVERT_CREATED_SUCCESS) {
    return { ...state, data: [action.payload, ...state.data] };
  }
  return state;
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return {
      isLoading: false,
      error: action.payload
    }
  }
  if (/_REQUEST$/.test(action.type)) {
    return {
      error: null,
      isLoading: true,
    };
  }
  if (/_SUCCESS$/.test(action.type)) {
    return {
      error: null,
      isLoading: false,
    };
  }
  if (action.type === UI_RESET_ERROR) {
    return {
      ...state,
      error: null,
    };
  }
  return state;
}