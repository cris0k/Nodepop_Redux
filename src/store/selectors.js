export const getIsLogged = state => state.auth;

export const getStateAdverts = state => state.adverts.data;

export const areAdvertsLoaded = state=> state.adverts.areLoaded;

// export const getAdvert = (state, advertId) =>
//   getStateAdverts(state).find(advert  => advert.id.toString() === advertId);

export const getDetails =  advertId => state =>
 getStateAdverts(state).find(advert  => advert.id.toString() === advertId);

  export const getUi = state => state.ui;