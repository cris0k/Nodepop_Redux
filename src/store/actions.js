import { areAdvertsLoaded, getDetails } from './selectors';
import { 
  AUTH_LOGIN_SUCCESS, 
  AUTH_LOGOUT, 
  AUTH_LOGIN_REQUEST, 
  AUTH_LOGIN_FAILURE, 
  UI_RESET_ERROR, 
  ADVERTS_LOADED_REQUEST, 
  ADVERTS_LOADED_SUCCESS, 
  ADVERTS_LOADED_FAILURE, 
  DETAILS_LOADED_REQUEST, 
  DETAILS_LOADED_SUCCESS,
  DETAILS_LOADED_FAILURE,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_SUCCESS,
  ADVERT_CREATED_FAILURE} 
  from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogin = credentials => {
  return async function(dispatch, getState, { api }){
    try {
      dispatch(authLoginRequest())
      await api.auth.login(credentials)
      dispatch(authLoginSuccess())

    } catch (error) {
      dispatch(authLoginFailure(error))
      
    }
  }
}

export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT,
});

export const authLogout = () => {
  return async function (dispatch, getState, { api }) {
    await api.auth.logout();
    dispatch(authLogoutSuccess());
  };
};

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST
  
});
export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});
export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true
});
export const advertsLoad = ()=>{
  return async function(dispatch,getState,{ api }){
    
    const areLoaded = areAdvertsLoaded(getState()) 
    if (areLoaded) return ;//if storedAdverts = true it'll get the adverts in state,else try/catch

    try {
      dispatch(advertsLoadedRequest())
      const adverts = await api.adverts.getAdverts()
      dispatch(advertsLoadedSuccess(adverts))
    } catch (error) {
      dispatch(advertsLoadedFailure(error))
    }
  }
}

export const detailsLoadedRequest = () => ({
  type: DETAILS_LOADED_REQUEST
  
});
export const detailsLoadedSuccess = advert => ({
  type: DETAILS_LOADED_SUCCESS,
  payload: advert,
});
export const detailsLoadedFailure = error => ({
  type: DETAILS_LOADED_FAILURE,
  payload: error,
  error: true
});

export const detailsLoad = advertId =>{
  return async function(dispatch,getState,{ api }){
    
    const isLoaded = getDetails(advertId)(getState()) 
    if (isLoaded) return ;

    try {
      dispatch(detailsLoadedRequest())
      const advert = await api.adverts.getAdvert(advertId)
      dispatch(detailsLoadedSuccess(advert))
    } catch (error) {
      dispatch(detailsLoadedFailure(error))
    }
  }
}

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});

export const advertCreatedRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreatedSuccess = advert => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
});

export const advertCreatedFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  payload: error,
  error: true,
});

export const advertCreate = advert => {
  return async function (dispatch, getState, { api }) {
    try {
      dispatch(advertCreatedRequest());
      const { id } = await api.adverts.createAdvert(advert);
      const createdAdvert = await api.adverts.getAdvert(id);
      dispatch(advertCreatedSuccess(createdAdvert));
      return createdAdvert;
    } catch (error) {
      dispatch(advertCreatedFailure(error));
    }
  };
};