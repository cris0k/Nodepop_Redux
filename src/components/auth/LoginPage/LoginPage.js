import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import LoginForm from './LoginForm';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { getUi } from '../../../store/selectors';

function LoginPage() {
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);
  const handleResetError = () => dispatch(uiResetError)

  const handleSubmit = async credentials => {

    await dispatch(authLogin(credentials))
    const from = location.state?.from?.pathname || '/';
    navigate(from, { replace: true });
     
  };

  return (
    <div>
      <h1>Log In</h1>
      <LoginForm onSubmit={handleSubmit} isLoading={isLoading} />
      { isLoading && <p>...login in nodepop </p>}
      {error && (
        <div onClick={handleResetError} style={{ color: 'red' }}>
          {error.message}
        </div> )}
      
    </div>
  );
}

export default LoginPage;
