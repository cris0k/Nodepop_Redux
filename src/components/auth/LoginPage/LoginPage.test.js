import {fireEvent, render, screen} from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router} from 'react-router-dom';
import { authLogin } from '../../../store/actions';
import { defaultState } from '../../../store/reducers';
import LoginPage from './LoginPage';

jest.mock('../../../store/actions');

describe('LoginPage', () => {
    const store = {
      getState: () => defaultState,
      dispatch: () => {},
      subscribe: () => {},
    };
  
    const renderComponent = () =>
      render(
        <Provider store={store}>
            <Router>
                <LoginPage />
            </Router>
        </Provider>,
      );
  
    test('snapshot', () => {
      const { container } = renderComponent();
      expect(container).toMatchSnapshot();
    });
  
    test('should dispatch authLogin action', () => {
      const credentials = {
        email:'cris', 
        password:'1234',
        remember: false
    }
      renderComponent();
  
      const emailInput = screen.getByPlaceholderText(/email/)
      const passwordInput = screen.getByPlaceholderText(/password/)
      const rememberInput = screen.getByRole('checkbox')
      const submitButton = screen.getByRole('button');

      fireEvent.change(emailInput, {target :{value: credentials.email} });
      fireEvent.change(passwordInput, {target :{value: credentials.password} });
      fireEvent.change(rememberInput,{target : { value :credentials.remember}})
  
      fireEvent.click(submitButton);
      expect(authLogin).toHaveBeenCalledWith(credentials);
    });
  });
