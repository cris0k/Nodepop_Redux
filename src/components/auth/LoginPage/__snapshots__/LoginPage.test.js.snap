// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div>
    <h1>
      Log In
    </h1>
    <form>
      <input
        name="email"
        placeholder="email"
        value=""
      />
      <input
        name="password"
        placeholder="password"
        type="password"
        value=""
      />
      <span>
        Remember
      </span>
      <input
        name="remember"
        type="checkbox"
      />
      <button
        disabled=""
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
